In [1]:
# To support both python 2 and python 3
from __future__ import division, print_function, unicode_literals

# Common imports
import numpy as np
import os

# to make this notebook's output stable across runs
def reset_graph(seed=42):
    tf.reset_default_graph()
    tf.set_random_seed(seed)
    np.random.seed(seed)

# To plot pretty figures
%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt
plt.rcParams['axes.labelsize'] = 14
plt.rcParams['xtick.labelsize'] = 12
plt.rcParams['ytick.labelsize'] = 12

# Where to save the figures
PROJECT_ROOT_DIR = "."
CHAPTER_ID = "deep"

def save_fig(fig_id, tight_layout=True):
    path = os.path.join(PROJECT_ROOT_DIR, "images", CHAPTER_ID, fig_id + ".png")
    print("Saving figure", fig_id)
    if tight_layout:
        plt.tight_layout()
    plt.savefig(path, format='png', dpi=300)

## Assignment 8.1 

Build a DNN with five hidden layers of 100 neurons each, He initialization, and the ELU activation function

In [34]:
import tensorflow as tf

n_inputs = 28 * 28
n_hidden1 = 100
n_hidden2 = 100
n_hidden3 = 100
n_hidden4 = 100
n_hidden5 = 100
n_outputs = 5

he_init = tf.contrib.layers.variance_scaling_initializer()

In [35]:
from functools import partial

dense_layer = partial(tf.layers.dense, activation=tf.nn.elu, kernel_initializer=he_init)

In [36]:
reset_graph()

X = tf.placeholder(tf.float32, shape=(None, n_inputs), name='X')
y = tf.placeholder(tf.int64, shape=(None), name='y')

with tf.name_scope('dnn'):
    hidden1 = dense_layer(X, n_hidden1, name='hidden1')
    hidden2 = dense_layer(hidden1, n_hidden2, name='hidden2')
    hidden3 = dense_layer(hidden2, n_hidden3, name='hidden3')
    hidden4 = dense_layer(hidden3, n_hidden4, name='hidden4')
    hidden5 = dense_layer(hidden4, n_hidden5, name='hidden5')
    logits = dense_layer(hidden5, n_outputs, activation=None, name='outputs')
    
with tf.name_scope('loss'):
    xentropy = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=y, logits=logits)
    loss = tf.reduce_mean(xentropy, name='loss')

## Assignment 8.2

Using Adam optimization and early stopping, try training it on MNIST but only on digits 0 to 4, as we will use transfer learning for digits 5 to 9 in the next exercise. You will need a softmax output layer with five neurons, and as always make sure to save checkpoints at regular intervals and save the final model so you can reuse it later.

`Adam optimation`: adaptive moment estimation, combines the ideas of Momentum optimization and RMSProp.

`Momentum optimization`: keep track of an exponentially decaying average of past gradients.

`RMSProp`: keep track of an exponentially decaying of past squared gradients.

In [37]:
learning_rate = 0.001

with tf.name_scope('train'):
    optimizer = tf.train.AdamOptimizer(learning_rate)
    training_op = optimizer.minimize(loss)

In [38]:
with tf.name_scope('eval'):
    correct = tf.nn.in_top_k(logits, y, 1)
    accuracy = tf.reduce_mean(tf.cast(correct, tf.float32))

In [39]:
from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets('/tmp/data/')

Extracting /tmp/data/train-images-idx3-ubyte.gz
Extracting /tmp/data/train-labels-idx1-ubyte.gz
Extracting /tmp/data/t10k-images-idx3-ubyte.gz
Extracting /tmp/data/t10k-labels-idx1-ubyte.gz


In [40]:
X_train = mnist.train.images[mnist.train.labels < 5]
y_train = mnist.train.labels[mnist.train.labels < 5]
X_test = mnist.test.images[mnist.test.labels < 5]
y_test = mnist.test.labels[mnist.test.labels < 5]

In [41]:
def shuffle_split(X, y, n_batches):
    np.random.seed(seed=42)
    rnd_idx = np.random.permutation(len(X))
    for i_idx in np.array_split(rnd_idx, n_batches):
        X_batch = X[i_idx]
        y_batch = y[i_idx]
        yield X_batch, y_batch

In [42]:
n_epochs = 50
batch_size = 50
n_batches = len(X_train) // batch_size
best_loss = float('inf')
patience = 2
cnt_patience = 0

init = tf.global_variables_initializer()

with tf.Session() as sess:
    init.run()
    for epoch in range(n_epochs):
        for X_batch, y_batch in shuffle_split(X_train, y_train, n_batches):
            sess.run([training_op, loss], feed_dict={X: X_batch, y: y_batch})
        accuracy_train = accuracy.eval(feed_dict={X: X_batch, y: y_batch})
        accuracy_test = accuracy.eval(feed_dict={X: X_test, y: y_test})
        loss_test = loss.eval(feed_dict={X: X_test, y: y_test})
        print(epoch, 'loss', loss_test, 'accuracy_train:', accuracy_train, 'accuracy_test:', accuracy_test)
        if loss_test < best_loss:
            best_loss = loss_test
        else:
            cnt_patience += 1
            if cnt_patience > patience:
                'Early stopping!'
                break

0 loss 0.04894622 accuracy_train: 1.0 accuracy_test: 0.9820977
1 loss 0.04120996 accuracy_train: 0.98 accuracy_test: 0.98715705
2 loss 0.03728239 accuracy_train: 1.0 accuracy_test: 0.98813
3 loss 0.030111544 accuracy_train: 1.0 accuracy_test: 0.9910488
4 loss 0.026404124 accuracy_train: 1.0 accuracy_test: 0.9918272
5 loss 0.0332601 accuracy_train: 1.0 accuracy_test: 0.99065965
6 loss 0.030299986 accuracy_train: 1.0 accuracy_test: 0.9918272
7 loss 0.02770311 accuracy_train: 1.0 accuracy_test: 0.99143803


Alternatively we can use `tf.keras.callbacks.EarlyStopping`.

## Assignment 8.3

Tune the hyperparameters using cross-validation and see what precision you can achieve.

### Order to tune hyperparameters in Neural Networks

Answers from [stackoverflow](https://stackoverflow.com/questions/37467647/in-what-order-should-we-tune-hyperparameters-in-neural-networks):

**My general order is:**

- Batch size, as it will largely affect the training time of future experiments.

- Architecture of the network:

    - Number of neurons in the network
    - Number of layers

- Rest (dropout, L2 reg, etc.)

**Dependencies:**

I'd assume that the optimal values of

- learning rate and batch size
- learning rate and number of neurons
- number of neurons and number of layers

strongly depend on each other. I am not an expert on that field though.

**As for your hyperparameters:**

- For the Adam optimizer: "Recommended values in the paper are eps = 1e-8, beta1 = 0.9, beta2 = 0.999." (source)
- For the learning rate with Adam and RMSProp, I found values around 0.001 to be optimal for most problems.
- As an alternative to Adam, you can also use RMSProp, which reduces the memory footprint by up to 33%. See this answer for more details.
- You could also tune the initial weight values (see All you need is a good init). Although, the Xavier initializer seems to be a good way to prevent having to tune the weight inits.
- I don't tune the number of iterations / epochs as a hyperparameter. I train the net until its validation error converges. However, I give each run a time budget.

### Create custom estimator with scikit learn

The following refers to [scikit learn documentation](http://scikit-learn.org/dev/developers/contributing.html#rolling-your-own-estimator) and a blog article from [Daniel Hnyk](http://danielhnyk.cz/creating-your-own-estimator-scikit-learn/).

**Estimator types**

Some common functionality depends on the kind of estimator passed. For different tasks, you can choose:

- ClassifierMixin
- RegressorMixin
- ClusterMixin

**get_params and set_params**

All estimators must have `get_params` and `set_params` functions. They are inherited from `BaseEstimator`.

**Pipeline compatibility**

For an estimator to be usable together with `pipeline.Pipeline`in any but the last step, it needs to provide a `fit` or `fit_transform` function. 

To be able to evaluate the pipeline on any data but the training set, it also needs to provide a `transform` function.


In [27]:
from sklearn.base import BaseEstimator, ClassifierMixin
import tensorflow as tf
from functools import partial
import numpy as np


class DnnClassifier(BaseEstimator, ClassifierMixin):
    
    
    def __init__(self, 
                 batch_size=50, 
                 n_neuron=100):
        
        self.batch_size = batch_size
        self.n_neuron = n_neuron
        
    def reset_graph(self, seed=42):
        tf.reset_default_graph()
        tf.set_random_seed(seed)
        np.random.seed(seed)
        
    def _build_graph(self, n_inputs, n_outputs):
        
        self.reset_graph()
        
        X = tf.placeholder(tf.float32, shape=(None, n_inputs), name='X')
        y = tf.placeholder(tf.int64, shape=(None), name='y')
        
        he_init = tf.contrib.layers.variance_scaling_initializer()
        
        dense_layer = partial(tf.layers.dense, activation=tf.nn.elu, kernel_initializer=he_init)

        with tf.name_scope('dnn'):
            hidden1 = dense_layer(X, self.n_neuron, name='hidden1')
            hidden2 = dense_layer(hidden1, self.n_neuron, name='hidden2')
            hidden3 = dense_layer(hidden2, self.n_neuron, name='hidden3')
            hidden4 = dense_layer(hidden3, self.n_neuron, name='hidden4')
            hidden5 = dense_layer(hidden4, self.n_neuron, name='hidden5')
            logits = dense_layer(hidden5, n_outputs, activation=None, name='outputs')
            
        with tf.name_scope('softmax'):
            y_proba = tf.nn.softmax(logits, axis=1, name='y_proba')
    
        with tf.name_scope('loss'):
            xentropy = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=y, logits=logits)
            loss = tf.reduce_mean(xentropy, name='loss')
        
        learning_rate = 0.001

        with tf.name_scope('train'):
            optimizer = tf.train.AdamOptimizer(learning_rate)
            training_op = optimizer.minimize(loss)
            
        with tf.name_scope('eval'):
            correct = tf.nn.in_top_k(logits, y, 1)
            accuracy = tf.reduce_mean(tf.cast(correct, tf.float32))
            
        self._training_op = training_op
        self._accuracy = accuracy
        self._X = X
        self._y = y
        self._logits = logits
        self._y_proba = y_proba
        
    def shuffle_split(self, X, y, n_batches):
        np.random.seed(seed=42)
        rnd_idx = np.random.permutation(len(X))
        for i_idx in np.array_split(rnd_idx, n_batches):
            X_batch = X[i_idx]
            y_batch = y[i_idx]
            yield X_batch, y_batch
        
    def fit(self, X, y, n_epochs=5, X_valid=None, y_valid=None):
        self.n_batches = len(X) // self.batch_size
        
        n_inputs = X.shape[1]
        self.n_outputs = len(np.unique(y))
        self.classes_ = np.unique(y)
        
        self._build_graph(n_inputs, self.n_outputs)
        
        init = tf.global_variables_initializer()

        self._session = tf.Session()
        with self._session.as_default() as sess:
            init.run()
            for epoch in range(n_epochs):
                for X_batch, y_batch in self.shuffle_split(X, y, self.n_batches):
                    feed_dict = {self._X: X_batch, self._y: y_batch}
                    sess.run([self._training_op], feed_dict=feed_dict)
                accuracy_train = self._accuracy.eval(feed_dict=feed_dict)
                if X_valid is not None and y_valid is not None:
                    accuracy_valid = self._accuracy.eval(feed_dict={self._X: X_valid, self._y: y_valid})
                    print('epoch:', epoch, 'accuracy_valid:', accuracy_valid*100)
                else:
                    print('epoch:', epoch, 'accuracy_train:', accuracy_train*100)
                
                
        return self
    
    def predict_proba(self, X):
        with self._session.as_default() as sess:
            y_proba = sess.run([self._y_proba], feed_dict={self._X: X})
            return np.array(y_proba).reshape((len(X), self.n_outputs))
        
    def predict(self, X):
        class_indices = np.argmax(self.predict_proba(X), axis=1)
        return np.array([[self.classes_[class_index]]
                         for class_index in class_indices], np.int32)

In [20]:
from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets('/tmp/data/')

X_train = mnist.train.images
y_train = mnist.train.labels
X_valid = mnist.test.images
y_valid = mnist.test.labels

Extracting /tmp/data/train-images-idx3-ubyte.gz
Extracting /tmp/data/train-labels-idx1-ubyte.gz
Extracting /tmp/data/t10k-images-idx3-ubyte.gz
Extracting /tmp/data/t10k-labels-idx1-ubyte.gz


In [22]:
dnn_clf = DnnClassifier()
dnn_clf.fit(X_train, y_train, 20, X_valid, y_valid)

epoch: 0 accuracy_valid: 94.73000168800354
epoch: 1 accuracy_valid: 96.13000154495239
epoch: 2 accuracy_valid: 96.670001745224
epoch: 3 accuracy_valid: 96.5399980545044
epoch: 4 accuracy_valid: 96.02000117301941
epoch: 5 accuracy_valid: 96.42999768257141
epoch: 6 accuracy_valid: 97.39999771118164
epoch: 7 accuracy_valid: 97.40999937057495
epoch: 8 accuracy_valid: 96.74999713897705
epoch: 9 accuracy_valid: 97.49000072479248
epoch: 10 accuracy_valid: 97.39000201225281
epoch: 11 accuracy_valid: 97.47999906539917
epoch: 12 accuracy_valid: 96.99000120162964
epoch: 13 accuracy_valid: 97.24000096321106
epoch: 14 accuracy_valid: 97.72999882698059
epoch: 15 accuracy_valid: 97.680002450943
epoch: 16 accuracy_valid: 97.39000201225281
epoch: 17 accuracy_valid: 97.57999777793884
epoch: 18 accuracy_valid: 97.39999771118164
epoch: 19 accuracy_valid: 97.53000140190125


DnnClassifier(batch_size=50, n_neuron=100)

Accuracy on the valid set is 97.53%, which is no bad. So let's try to tune the n_neuron and batch_size to see if it can be improved.

In [29]:
from sklearn.grid_search import GridSearchCV

params_grid = [{'batch_size': [20, 50]}, {'n_neuron': [50, 100]}]
gs = GridSearchCV(DnnClassifier(), params_grid, 
                 fit_params={'n_epochs': 20, 'X_valid': X_valid, 'y_valid': y_valid})

gs.fit(X_train, y_train)

gs.best_params_

epoch: 0 accuracy_valid: 95.46999931335449
epoch: 1 accuracy_valid: 95.56000232696533
epoch: 2 accuracy_valid: 95.52000164985657
epoch: 3 accuracy_valid: 96.07999920845032
epoch: 4 accuracy_valid: 96.64999842643738
epoch: 5 accuracy_valid: 96.32999897003174
epoch: 6 accuracy_valid: 96.85999751091003
epoch: 7 accuracy_valid: 96.45000100135803
epoch: 8 accuracy_valid: 96.99000120162964
epoch: 9 accuracy_valid: 96.61999940872192
epoch: 10 accuracy_valid: 96.61999940872192
epoch: 11 accuracy_valid: 96.60000205039978
epoch: 12 accuracy_valid: 96.90999984741211
epoch: 13 accuracy_valid: 96.96999788284302
epoch: 14 accuracy_valid: 95.98000049591064
epoch: 15 accuracy_valid: 96.27000093460083
epoch: 16 accuracy_valid: 97.00999855995178
epoch: 17 accuracy_valid: 97.21999764442444
epoch: 18 accuracy_valid: 97.06000089645386
epoch: 19 accuracy_valid: 96.93999886512756
epoch: 0 accuracy_valid: 93.65000128746033
epoch: 1 accuracy_valid: 95.63000202178955
epoch: 2 accuracy_valid: 95.63000202178955
e

epoch: 10 accuracy_valid: 97.10000157356262
epoch: 11 accuracy_valid: 96.78999781608582
epoch: 12 accuracy_valid: 96.46000266075134
epoch: 13 accuracy_valid: 96.86999917030334
epoch: 14 accuracy_valid: 96.7199981212616
epoch: 15 accuracy_valid: 97.2100019454956
epoch: 16 accuracy_valid: 96.88000082969666
epoch: 17 accuracy_valid: 97.28999733924866
epoch: 18 accuracy_valid: 96.85999751091003
epoch: 19 accuracy_valid: 96.64999842643738
epoch: 0 accuracy_valid: 95.06000280380249
epoch: 1 accuracy_valid: 95.81999778747559
epoch: 2 accuracy_valid: 95.75999975204468
epoch: 3 accuracy_valid: 96.17999792098999
epoch: 4 accuracy_valid: 96.10000252723694
epoch: 5 accuracy_valid: 96.06000185012817
epoch: 6 accuracy_valid: 96.61999940872192
epoch: 7 accuracy_valid: 96.34000062942505
epoch: 8 accuracy_valid: 96.60999774932861
epoch: 9 accuracy_valid: 96.64000272750854
epoch: 10 accuracy_valid: 96.8999981880188
epoch: 11 accuracy_valid: 97.28000164031982
epoch: 12 accuracy_valid: 96.60000205039978
e

{'batch_size': 50}

In [33]:
from sklearn.metrics import accuracy_score

y_valid_pred = gs.predict(X_valid)
print(accuracy_score(y_valid, y_valid_pred))

0.9753


The accuracy is not improved after tuning. It might be because that the `fit` method doesn't implement early stopping yet.